File: algorithms/cpp/33.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution
{
public:
    int search(vector<int>& nums, int target)
    {
        int l = 0, r = nums.size()-1;
        while ( l <= r )
        {
            int m = (l+r)/2;
            if ( nums[m] == target ) return m;
            else if ( target < nums[m] )
            {
                if ( nums[l] <= target || nums[m] < nums[l] )
                    r = m-1;
                else
                    l = m+1;
            }
            else
            {
                if ( nums[l] > target || nums[m] >= nums[l] )
                    l = m+1;
                else
                    r = m-1;
            }
        }
        return -1;
    }
};

int main()
{
    return 0;
}

File: algorithms/rust/33.rs
impl Solution {
    #[allow(dead_code)]
    pub fn search(nums: Vec<i32>, target: i32) -> i32 {
        let (mut left, mut right) = (0_i32, nums.len() as i32 - 1);
        while left <= right {
            let mid = left + (right - left) / 2;
            let mid_val = nums[mid as usize];
            if mid_val == target {
                return mid;
            }

            if (target > mid_val || target < nums[0]) && mid_val >= nums[0]
                || target < nums[0] && target > mid_val
            {
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }
        -1
    }
}

struct Solution;

fn main() {
    println!("Hello, world!");
}
